storage_controller/src/service.rs

```diff
@@ -3734,6 +3735,60 @@ impl Service {
         Ok(())
     }
 
+    pub(crate) async fn tenant_timeline_download_heatmap_layers(
+        &self,
+        tenant_shard_id: TenantShardId,
+        timeline_id: TimelineId,
+        concurrency: Option<usize>,
+    ) -> Result<(), ApiError> {
+        let _tenant_lock = trace_shared_lock(
+            &self.tenant_op_locks,
+            tenant_shard_id.tenant_id,
+            TenantOperations::DownloadHeatmapLayers,
+        )
+        .await;
+
+        let targets = {
```

Comment:
Bit surprised this doesn't use `tenant_for_shards_api` or some similar helper, did it not fit the interface?  (if so let's have a comment to say why it hand rolls this)

Reply:
Oh, it's to make it automagically work as an all-shards or one-shard action, right?  Hmm.  Maybe we could just teach the helper the same trick?

Reply:
Switched to using `tenant_for_shards_api`: https://github.com/neondatabase/neon/pull/10788/commits/8e81c9f2d7da44d1e49291356cbd849d8e0eee63